packages/marshal/marshal.js

```diff
@@ -1,7 +1,70 @@
-import harden from '@agoric/harden';
+// @ts-check
+import rawHarden from '@agoric/harden';
 import Nat from '@agoric/nat';
 import { HandledPromise } from '@agoric/eventual-send';
 
+const harden = /** @type {<T>(x: T) => T} */ (rawHarden);
+
+const presenceToInterface = new WeakMap();
+
+/**
+ * Simple semantics, just tell what interface (or undefined) a presence has.
+ *
+ * @param {*} maybePresence the value to check
+ * @returns {*} the interface, or undefined if not a Presence
+ */
+export function getInterfaceOf(maybePresence) {
+  return presenceToInterface.get(maybePresence);
+}
+
+/**
+ * Do a deep copy of the object, handling Proxies and recursion.
+ * The resulting copy is guaranteed to be pure data, as well as hardened.
+ * Such a hardened, pure copy cannot be used as a communications path.
+ *
+ * @template T
+ * @param {T} val input value
+ * @returns {T} pure, hardened copy
+ */
+function pureCopy(val, already = new WeakMap()) {
+  const valType = typeof val;
```

Comment:
This doesn't enforce that the original or the result is deeply pass-by-copy. For example, if it encounters a presence, it won't reject it or preserve the identity. It doesn't check the `getPrototypeOf`. It doesn't reject symbol-named properties.

Rather than doing a `typeof`, consider doing a `passStyleOf`.

Reply:
Done.

The thing I don't like about `passStyleOf` is that it insists on having its argument be frozen.  Is there any way we can lift this restriction so that `pureCopy` only has to harden the copy instead of the original object?

Reply:
(Not for all of `passStyleOf`, but just for the one used by `pureCopy`.)